openscenario/map_fragment/src/intersection.cpp

```diff
@@ -0,0 +1,254 @@
+// Copyright 2015 TIER IV, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map_fragment/map_fragment.hpp>
+#include <map_fragment/road_segments/generate_lanelet_map.hpp>
+#include <rclcpp/rclcpp.hpp>
+
+enum IntersectionLeg { LEFT_LEG = 0, BOTTOM_LEG = 1, RIGHT_LEG = 2, TOP_LEG = 3 };
+
+auto getLegOrientation(IntersectionLeg leg) -> double { return leg * M_PI_2; }
+
+enum ConnectionType { LEFT_TURN = -1, FORWARD = 2, RIGHT_TURN = 1 };
+
+auto getConnectedLeg(IntersectionLeg leg, ConnectionType connection_type) -> IntersectionLeg
+{
+  return static_cast<IntersectionLeg>(((leg + connection_type) + 4) % 4);
+}
+
+auto main(const int argc, char const * const * const argv) -> int
+try {
+  using namespace map_fragment;
+  using namespace map_fragment::road_segments;
+
+  rclcpp::init(argc, argv);
+
+  auto node = rclcpp::Node(std::filesystem::path(argv[0]).stem());
+
+  const auto lane_width = [&]() {
+    node.declare_parameter("lane_width", default_value::width);
+    return node.get_parameter("lane_width").as_double();
+  }();
+
+  const auto leg_length = [&]() {
+    node.declare_parameter("leg_length", default_value::length);
+    return node.get_parameter("leg_length").as_double();
+  }();
+
+  const auto turn_radius = [&]() {
+    node.declare_parameter("turn_radius", default_value::turn_radius);
+    return node.get_parameter("turn_radius").as_double();
+  }();
+
+  const auto resolution = [&]() {
+    node.declare_parameter("resolution", default_value::resolution);
+    return node.get_parameter("resolution").as_int();
+  }();
+
+  const auto number_of_lanes_going_left = [&]() {
+    node.declare_parameter("number_of_lanes_going_left", 0);
+    return node.get_parameter("number_of_lanes_going_left").as_int();
+  }();
+
+  const auto number_of_lanes_going_left_and_straight = [&]() {
+    node.declare_parameter("number_of_lanes_going_left_and_straight", 0);
+    return node.get_parameter("number_of_lanes_going_left_and_straight").as_int();
+  }();
+
+  const auto number_of_lanes_going_straight = [&]() {
+    node.declare_parameter("number_of_lanes_going_straight", 0);
+    return node.get_parameter("number_of_lanes_going_straight").as_int();
+  }();
+
+  const auto number_of_lanes_going_in_all_directions = [&]() {
+    node.declare_parameter("number_of_lanes_going_in_all_directions", 0);
+    return node.get_parameter("number_of_lanes_going_in_all_directions").as_int();
+  }();
```

Comment:
If the value given to `number_of_lanes_going_in_all_directions` exceeds 6, a malformed intersection is generated (This threshold may vary depending on the value given to `turn_radius`.) This problem is unacceptable because 8-lane intersection actually exists. In addition, please make sure that the proper intersection is generated even if a value too large, such as 100, is given. This is because the inability to generate the proper intersection in such a case implies that there is a code somewhere that is making incorrect assumptions.

Reply:
You are correct, the malformed intersection is due to too small value of `turn_radius` parameter. If [the newly added check](https://github.com/tier4/scenario_simulator_v2/pull/1124/files#diff-9f581456265172861eea619b8d2400cf8775f29cbbf37b15b9b6c5fa0fa54ccbR102-R109) fails, it will give hints what minimal value should be provided for given configuration.

I wasn't able to load a 100-lane intersection into Scenario Editor to validate it (or perhaps I was just too impatient to wait for so long), but here is a 50-lane intersection (minimal turn radius in this case is 500.0):

```
ros2 run map_fragment intersection --ros-args \
-p number_of_lanes_going_in_all_directions:=50 \
-p turn_radius:=600.0
```

![image](https://github.com/tier4/scenario_simulator_v2/assets/2529619/50cc38d9-7188-4d79-86d8-5c525652c9f5)

It looks to me like a microcontroller.